src/dwarf.rs

```diff
@@ -952,11 +1005,114 @@ impl DwarfResolver {
     ///
     /// This function is pretty much the same as `find_line_as_ref()`
     /// except returning a copies of `String` instead of `&str`.
+    #[allow(dead_code)]
     pub fn find_line(&self, address: u64) -> Option<(String, String, usize)> {
         let (dir, file, line_no) = self.find_line_as_ref(address)?;
         Some((String::from(dir), String::from(file), line_no))
     }
 
+    /// Extract the symbol information from DWARf if having not did it
+    /// before.
+    fn ensure_debug_info_syms(&self) -> Result<(), Error> {
+        if self.enable_debug_info_syms {
+            let mut dis_ref = self.debug_info_syms.borrow_mut();
+            if dis_ref.is_some() {
+                return Ok(());
+            }
+            let mut debug_info_syms = debug_info_parse_symbols(&self.parser, None, 1)?;
+            debug_info_syms.sort_by_key(|v: &DWSymInfo| -> &str { v.name });
+            *dis_ref = Some(unsafe { mem::transmute(debug_info_syms) });
+        }
+        Ok(())
+    }
+
+    /// Find the address of a symbol from DWARF.
+    ///
+    /// # Arguments
+    ///
+    /// * `name` - is the symbol name to find.
+    /// * `ctx` - is the context giving additional parameters.
+    pub fn find_address(&self, name: &str, ctx: &FindAddrContext) -> Result<SymbolInfo, Error> {
+        if let SymbolType::Variable = ctx.sym_type {
+            return Err(Error::new(ErrorKind::Unsupported, "Not implemented"));
+        }
+        let r = self.parser.find_address(name, ctx);
+        if r.is_ok() {
+            return r;
+        }
+
+        self.ensure_debug_info_syms()?;
+        let dis_ref = self.debug_info_syms.borrow();
+        let debug_info_syms = dis_ref.as_ref().unwrap();
+        match debug_info_syms.binary_search_by_key(&name.to_string(), |v| v.name.to_string()) {
+            Ok(idx) => {
+                let DWSymInfo {
+                    address,
+                    size,
+                    sym_type,
+                    ..
+                } = debug_info_syms[idx];
+                Ok(SymbolInfo {
+                    name: name.to_string(),
+                    address,
+                    size,
+                    sym_type,
+                    ..Default::default()
+                })
+            }
+            Err(_) => Err(Error::new(ErrorKind::NotFound, "symbol not found")),
+        }
+    }
+
+    /// Find the address of symbols matching a pattern from DWARF.
+    ///
+    /// #Arguments
+    ///
+    /// * `pattern` - is a regex pattern to match symbols.
+    /// * `ctx` - is the context giving additional parameters.
+    ///
+    /// Return a list of symbols including addresses and other information.
+    pub fn find_addrs_regex(&self, pattern: &str, ctx: &FindAddrContext) -> Vec<SymbolInfo> {
+        if let SymbolType::Variable = ctx.sym_type {
+            return vec![];
```

Comment:
you return error in this case for find_address, why the difference?

Reply:
It should be the same.  I will change the return type to Result<Vec<SymbolInfo>, Error>.